File: abilities-common/src/main/java/io/github/solusmods/eternalcore/abilities/api/AbilityEvents.java
package io.github.solusmods.eternalcore.abilities.api;

import dev.architectury.event.Event;
import dev.architectury.event.EventFactory;
import dev.architectury.event.EventResult;
import io.github.solusmods.eternalcore.network.api.util.Changeable;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.world.damagesource.DamageSource;
import net.minecraft.world.entity.LivingEntity;

public interface AbilityEvents {
    Event<UnlockAbilityEvent> UNLOCK_ABILITY = EventFactory.createEventResult();
    Event<RemoveAbilityEvent> REMOVE_ABILITY = EventFactory.createEventResult();
    Event<AbilityActivationEvent> ACTIVATE_ABILITY = EventFactory.createEventResult();
    Event<AbilityReleaseEvent> RELEASE_ABILITY = EventFactory.createEventResult();
    Event<AbilityToggleEvent> TOGGLE_ABILITY = EventFactory.createEventResult();
    Event<AbilityScrollEvent> ABILITY_SCROLL = EventFactory.createEventResult();
    Event<AbilityScrollClientEvent> ABILITY_SCROLL_CLIENT = EventFactory.createEventResult();
    Event<AbilityTickEvent> ABILITY_PRE_TICK = EventFactory.createEventResult();
    Event<AbilityPostTickEvent> ABILITY_POST_TICK = EventFactory.createLoop();
    Event<AbilityUpdateCooldownEvent> ABILITY_UPDATE_COOLDOWN = EventFactory.createEventResult();
    Event<AbilityDamageCalculationEvent> ABILITY_DAMAGE_PRE_CALCULATION = EventFactory.createEventResult();
    Event<AbilityDamageCalculationEvent> ABILITY_DAMAGE_CALCULATION = EventFactory.createEventResult();
    Event<AbilityDamageCalculationEvent> ABILITY_DAMAGE_POST_CALCULATION = EventFactory.createEventResult();


    @FunctionalInterface
    interface UnlockAbilityEvent {
        EventResult unlockAbility(AbilityInstance instance, LivingEntity owner, Changeable<MutableComponent> unlockMessage);
    }

    @FunctionalInterface
    interface RemoveAbilityEvent {
        EventResult removeAbility(AbilityInstance instance, LivingEntity owner, Changeable<MutableComponent> forgetMessage);
    }

    @FunctionalInterface
    interface AbilityActivationEvent {
        EventResult activateAbility(Changeable<AbilityInstance> abilityInstance, LivingEntity owner, int keyNumber, int mode);
    }

    @FunctionalInterface
    interface AbilityReleaseEvent {
        EventResult releaseAbility(Changeable<AbilityInstance> abilityInstance, LivingEntity owner, int keyNumber, int mode, int heldTicks);
    }

    @FunctionalInterface
    interface AbilityToggleEvent {
        EventResult toggleAbility(Changeable<AbilityInstance> abilityInstance, LivingEntity owner);
    }

    @FunctionalInterface
    interface AbilityScrollEvent {
        EventResult scroll(Changeable<AbilityInstance> abilityInstance, LivingEntity owner, Changeable<Double> delta);
    }

    @FunctionalInterface
    interface AbilityScrollClientEvent {
        EventResult scroll(AbilityInstance instance, LivingEntity owner, double delta);
    }

    @FunctionalInterface
    interface AbilityTickEvent {
        EventResult tick(AbilityInstance instance, LivingEntity owner);
    }

    @FunctionalInterface
    interface AbilityPostTickEvent {
        void tick(AbilityInstance instance, LivingEntity owner);
    }

    @FunctionalInterface
    interface AbilityUpdateCooldownEvent {
        EventResult cooldown(AbilityInstance instance, LivingEntity owner, int currentCooldown, int mode);
    }

    @FunctionalInterface
    interface AbilityDamageCalculationEvent {
        EventResult calculate(Abilities storage, LivingEntity entity, DamageSource source, Changeable<Float> amount);
    }
}

File: abilities-common/src/main/kotlin/io/github/solusmods/eternalcore/abilities/api/AbilityEvents.kt
package io.github.solusmods.eternalcore.abilities.api

import dev.architectury.event.Event
import dev.architectury.event.EventFactory
import dev.architectury.event.EventResult
import io.github.solusmods.eternalcore.network.api.util.Changeable
import net.minecraft.network.chat.MutableComponent
import net.minecraft.world.damagesource.DamageSource
import net.minecraft.world.entity.LivingEntity

interface AbilityEvents {
    fun interface UnlockAbilityEvent {
        fun unlockAbility(
            instance: AbilityInstance?,
            owner: LivingEntity?,
            unlockMessage: Changeable<MutableComponent?>?
        ): EventResult?
    }

    fun interface RemoveAbilityEvent {
        fun removeAbility(
            instance: AbilityInstance?,
            owner: LivingEntity?,
            forgetMessage: Changeable<MutableComponent?>?
        ): EventResult?
    }

    fun interface AbilityActivationEvent {
        fun activateAbility(
            abilityInstance: Changeable<AbilityInstance?>?,
            owner: LivingEntity?,
            keyNumber: Int,
            mode: Int
        ): EventResult?
    }

    fun interface AbilityReleaseEvent {
        fun releaseAbility(
            abilityInstance: Changeable<AbilityInstance?>?,
            owner: LivingEntity?,
            keyNumber: Int,
            mode: Int,
            heldTicks: Int
        ): EventResult?
    }

    fun interface AbilityToggleEvent {
        fun toggleAbility(abilityInstance: Changeable<AbilityInstance?>?, owner: LivingEntity?): EventResult?
    }

    fun interface AbilityScrollEvent {
        fun scroll(
            abilityInstance: Changeable<AbilityInstance?>?,
            owner: LivingEntity?,
            delta: Changeable<Double?>?
        ): EventResult?
    }

    fun interface AbilityScrollClientEvent {
        fun scroll(instance: AbilityInstance?, owner: LivingEntity?, delta: Double): EventResult?
    }

    fun interface AbilityTickEvent {
        fun tick(instance: AbilityInstance?, owner: LivingEntity?): EventResult?
    }

    fun interface AbilityPostTickEvent {
        fun tick(instance: AbilityInstance?, owner: LivingEntity?)
    }

    fun interface AbilityUpdateCooldownEvent {
        fun cooldown(instance: AbilityInstance?, owner: LivingEntity?, currentCooldown: Int, mode: Int): EventResult?
    }

    fun interface AbilityDamageCalculationEvent {
        fun calculate(
            storage: Abilities?,
            entity: LivingEntity?,
            source: DamageSource?,
            amount: Changeable<Float?>?
        ): EventResult?
    }

    companion object {
        val UNLOCK_ABILITY: Event<UnlockAbilityEvent?> = EventFactory.createEventResult<UnlockAbilityEvent?>()
        val REMOVE_ABILITY: Event<RemoveAbilityEvent?> = EventFactory.createEventResult<RemoveAbilityEvent?>()
        val ACTIVATE_ABILITY: Event<AbilityActivationEvent?> = EventFactory.createEventResult<AbilityActivationEvent?>()
        val RELEASE_ABILITY: Event<AbilityReleaseEvent?> = EventFactory.createEventResult<AbilityReleaseEvent?>()
        val TOGGLE_ABILITY: Event<AbilityToggleEvent?> = EventFactory.createEventResult<AbilityToggleEvent?>()
        val ABILITY_SCROLL: Event<AbilityScrollEvent?> = EventFactory.createEventResult<AbilityScrollEvent?>()
        val ABILITY_SCROLL_CLIENT: Event<AbilityScrollClientEvent?> =
            EventFactory.createEventResult<AbilityScrollClientEvent?>()
        val ABILITY_PRE_TICK: Event<AbilityTickEvent?> = EventFactory.createEventResult<AbilityTickEvent?>()
        val ABILITY_POST_TICK: Event<AbilityPostTickEvent?> = EventFactory.createLoop<AbilityPostTickEvent?>()
        val ABILITY_UPDATE_COOLDOWN: Event<AbilityUpdateCooldownEvent?> =
            EventFactory.createEventResult<AbilityUpdateCooldownEvent?>()
        val ABILITY_DAMAGE_PRE_CALCULATION: Event<AbilityDamageCalculationEvent?> =
            EventFactory.createEventResult<AbilityDamageCalculationEvent?>()
        val ABILITY_DAMAGE_CALCULATION: Event<AbilityDamageCalculationEvent?> =
            EventFactory.createEventResult<AbilityDamageCalculationEvent?>()
        val ABILITY_DAMAGE_POST_CALCULATION: Event<AbilityDamageCalculationEvent?> =
            EventFactory.createEventResult<AbilityDamageCalculationEvent?>()
    }
}
